BadVersionException

Hey, guys. I keep getting this exception. I am using 0.87.1 version of Realm(and I use Realm as a singleton).  

My situation looks like this: there is navigation button which starts a fragment, which makes an async query for objects in onStart() method. This exception rises when user spams this navigation button or just makes fast transactions between fragments.

```java.lang.Error: io.realm.internal.async.BadVersionException: std::exception in io_realm_internal_TableQuery.cpp line 1262
E/AndroidRuntime:     at io.realm.RealmQuery$3.call(RealmQuery.java:1491)
E/AndroidRuntime:     at io.realm.RealmQuery$3.call(RealmQuery.java:1467)
E/AndroidRuntime:     at io.realm.internal.async.BgPriorityCallable.call(BgPriorityCallable.java:36)
E/AndroidRuntime:     at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:305)
E/AndroidRuntime:     at java.util.concurrent.FutureTask.run(FutureTask.java:137)
E/AndroidRuntime:     at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1076)
E/AndroidRuntime:     at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:569)
E/AndroidRuntime:     at java.lang.Thread.run(Thread.java:856)
E/AndroidRuntime:  Caused by: io.realm.internal.async.BadVersionException: std::exception in io_realm_internal_TableQuery.cpp line 1262
E/AndroidRuntime:     at io.realm.internal.TableQuery.nativeFindAllSortedWithHandover(Native Method)
E/AndroidRuntime:     at io.realm.internal.TableQuery.findAllSortedWithHandover(TableQuery.java:490)
E/AndroidRuntime:     at io.realm.RealmQuery$3.call(RealmQuery.java:1480)
E/AndroidRuntime:     at io.realm.RealmQuery$3.call(RealmQuery.java:1467) 
E/AndroidRuntime:     at io.realm.internal.async.BgPriorityCallable.call(BgPriorityCallable.java:36) 
E/AndroidRuntime:     at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:305) 
E/AndroidRuntime:     at java.util.concurrent.FutureTask.run(FutureTask.java:137) 
E/AndroidRuntime:     at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1076) 
E/AndroidRuntime:     at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:569) 
E/AndroidRuntime:     at java.lang.Thread.run(Thread.java:856)```

By the way, sometimes my during fast transactions between fragments my app crashes with SIGSEV  or just completely freezes without throwing ANR, dunno if it's connected to the BadVersion, but just in case.

I have similar error on version 0.87.4:
```
java.lang.Error: io.realm.internal.async.BadVersionException: std::exception in io_realm_internal_TableQuery.cpp line 1094
at io.realm.RealmQuery$5.call(RealmQuery.java:1847)
at io.realm.RealmQuery$5.call(RealmQuery.java:1817)
at io.realm.internal.async.BgPriorityCallable.call(BgPriorityCallable.java:36)
at java.util.concurrent.FutureTask.run(FutureTask.java:237)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
at java.lang.Thread.run(Thread.java:818)

Caused by: io.realm.internal.async.BadVersionException: std::exception in io_realm_internal_TableQuery.cpp line 1094
at io.realm.internal.TableQuery.nativeFindWithHandover(Native Method)
at io.realm.internal.TableQuery.findWithHandover(TableQuery.java:438)
at io.realm.RealmQuery$5.call(RealmQuery.java:1829)
at io.realm.RealmQuery$5.call(RealmQuery.java:1817) 
at io.realm.internal.async.BgPriorityCallable.call(BgPriorityCallable.java:36) 
at java.util.concurrent.FutureTask.run(FutureTask.java:237) 
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112) 
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587) 
at java.lang.Thread.run(Thread.java:818) 
```
Occurs randomly on app launch when it performs background network operations with writing results to Realm. On foreground activity I have listeners to Realm updates.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -2,6 +2,10 @@
 
 * Updated Realm Core to 0.97.1.
 
+### Bug fixes
+
+* Race condition causing BadVersionException when running multiple async writes and queries at the same time (#2021/#2391/#2417).
+
 ## 0.88.1
 
 ### Bug fixes

--- a/realm/realm-jni/src/io_realm_internal_TableQuery.cpp
+++ b/realm/realm-jni/src/io_realm_internal_TableQuery.cpp
@@ -1037,24 +1037,29 @@ JNIEXPORT jlong JNICALL Java_io_realm_internal_TableQuery_nativeFind(
     return -1;
 }
 
-std::unique_ptr<Query> getHandoverQuery (jlong bgSharedGroupPtr, jlong queryPtr, bool advanceToLatestVersion)
+// Returns a pointer to query on the worker SharedGroup or throw a BadVersion if the SharedGroup version required
+// for the handover is no longer available.
+std::unique_ptr<Query> handoverQueryToWorker(jlong bgSharedGroupPtr, jlong queryPtr, bool advanceToLatestVersion)
 {
     SharedGroup::Handover<Query> *handoverQueryPtr = HO(Query, queryPtr);
     std::unique_ptr<SharedGroup::Handover<Query>> handoverQuery(handoverQueryPtr);
 
-    SG(bgSharedGroupPtr)->end_read();
-
-    SharedGroup::VersionID currentVersion = SG(bgSharedGroupPtr)->get_version_of_current_transaction();
-    bool isDifferentVersions = (currentVersion != handoverQuery->version);
-    if (isDifferentVersions) {
-        SG(bgSharedGroupPtr)->begin_read(handoverQuery->version);
-    } else {
-        SG(bgSharedGroupPtr)->begin_read();
+    // The Handover object doesn't prevent a SharedGroup version from no longer being accessible. In rare
+    // cases this means that the version in the Handover object is invalid and Realm Core will throw a
+    // BadVersion as result.
+    realm::SharedGroup* sg = SG(bgSharedGroupPtr);
+    if (sg->get_transact_stage() != SharedGroup::transact_Reading) {
+        // if the SharedGroup is not in Read Transaction, we position it at the same version as the handover
+        sg->begin_read(handoverQuery->version);
+    } else if (sg->get_version_of_current_transaction() != handoverQuery->version) {
+        sg->end_read();
+        sg->begin_read(handoverQuery->version);
     }
 
-    std::unique_ptr<Query> query = SG(bgSharedGroupPtr)->import_from_handover(std::move(handoverQuery));
-    if (advanceToLatestVersion && isDifferentVersions) {
-        LangBindHelper::advance_read(*SG(bgSharedGroupPtr));
+    std::unique_ptr<Query> query = sg->import_from_handover(std::move(handoverQuery));
+
+    if (advanceToLatestVersion) {
+        LangBindHelper::advance_read(*sg);
     }
 
     return query;
@@ -1066,7 +1071,7 @@ JNIEXPORT jlong JNICALL Java_io_realm_internal_TableQuery_nativeFindWithHandover
 {
     TR_ENTER()
     try {
-        std::unique_ptr<Query> query = getHandoverQuery(bgSharedGroupPtr, queryPtr, false);
+        std::unique_ptr<Query> query = handoverQueryToWorker(bgSharedGroupPtr, queryPtr, false); // throws
         TableRef table = query->get_table();
 
         if (!QUERY_VALID(env, query.get())) {
@@ -1118,7 +1123,7 @@ JNIEXPORT jlong JNICALL Java_io_realm_internal_TableQuery_nativeFindAllWithHando
   {
       TR_ENTER()
       try {
-          std::unique_ptr<Query> query = getHandoverQuery(bgSharedGroupPtr, queryPtr, true);
+          std::unique_ptr<Query> query = handoverQueryToWorker(bgSharedGroupPtr, queryPtr, true); // throws
           return findAllWithHandover(env, bgSharedGroupPtr, std::move(query), start, end, limit);
       } CATCH_STD()
       return 0;
@@ -1146,26 +1151,35 @@ JNIEXPORT jlongArray JNICALL Java_io_realm_internal_TableQuery_nativeBatchUpdate
         std::vector<jlong> exported_handover_tableview_array(number_of_queries);
 
         // Step1: Position the shared group at the handover query version so we can import all queries
-
         // read the first query to determine the version we should use
         SharedGroup::Handover<Query> *handoverQueryPtr = HO(Query, handover_queries_pointer_array[0]);
         std::unique_ptr<SharedGroup::Handover<Query>> handoverQuery(handoverQueryPtr);
-        // position this shared group at the specified version
-        SG(bgSharedGroupPtr)->begin_read(handoverQuery->version);
+
+        // if the SharedGroup is not in Read Transaction, we position it at the same version as the handover
+        // The Handover object doesn't prevent a SharedGroup version from no longer being accessible. In rare
+        // cases this means that the version in the Handover object is invalid and Realm Core will throw a
+        // BadVersion as result.
+        realm::SharedGroup* sg = SG(bgSharedGroupPtr);
+        if (sg->get_transact_stage() != SharedGroup::transact_Reading) {
+            sg->begin_read(handoverQuery->version);
+        } else if (sg->get_version_of_current_transaction() != handoverQuery->version) {
+            sg->end_read();
+            sg->begin_read(handoverQuery->version);
+        }
 
         std::vector<std::unique_ptr<Query>> queries(number_of_queries);
 
         // import the first query
-        queries[0] = std::move(SG(bgSharedGroupPtr)->import_from_handover(std::move(handoverQuery)));
+        queries[0] = std::move(sg->import_from_handover(std::move(handoverQuery)));
 
         // import the rest of the queries
         for (size_t i = 1; i < number_of_queries; ++i) {
             std::unique_ptr<SharedGroup::Handover<Query>> handoverQuery(HO(Query, handover_queries_pointer_array[i]));
-            queries[i] = std::move(SG(bgSharedGroupPtr)->import_from_handover(std::move(handoverQuery)));
+            queries[i] = std::move(sg->import_from_handover(std::move(handoverQuery)));
         }
 
         // Step2: Bring the queries into the latest shared group version
-        LangBindHelper::advance_read(*SG(bgSharedGroupPtr));
+        LangBindHelper::advance_read(*sg);
 
         // Step3: Run & export the queries against the latest shared group
         for (size_t i = 0; i < number_of_queries; ++i) {
@@ -1246,7 +1260,7 @@ JNIEXPORT jlong JNICALL Java_io_realm_internal_TableQuery_nativeGetDistinctViewW
 {
     TR_ENTER()
     try {
-        std::unique_ptr<Query> query = getHandoverQuery(bgSharedGroupPtr, queryPtr, true);
+        std::unique_ptr<Query> query = handoverQueryToWorker(bgSharedGroupPtr, queryPtr, true); // throws
         return getDistinctViewWithHandover(env, bgSharedGroupPtr, std::move(query), columnIndex);
     } CATCH_STD()
     return 0;
@@ -1257,7 +1271,7 @@ JNIEXPORT jlong JNICALL Java_io_realm_internal_TableQuery_nativeFindAllSortedWit
   {
       TR_ENTER()
       try {
-          std::unique_ptr<Query> query = getHandoverQuery(bgSharedGroupPtr, queryPtr, true);
+          std::unique_ptr<Query> query = handoverQueryToWorker(bgSharedGroupPtr, queryPtr, true); // throws
           return findAllSortedWithHandover(env, bgSharedGroupPtr, std::move(query), start, end, limit, columnIndex, ascending);
       } CATCH_STD()
       return 0;
@@ -1269,7 +1283,7 @@ JNIEXPORT jlong JNICALL Java_io_realm_internal_TableQuery_nativeFindAllMultiSort
       TR_ENTER()
       try {
           // import the handover query pointer using the background SharedGroup
-          std::unique_ptr<Query> query = getHandoverQuery(bgSharedGroupPtr, queryPtr, true);
+          std::unique_ptr<Query> query = handoverQueryToWorker(bgSharedGroupPtr, queryPtr, true); // throws
           return findAllMultiSortedWithHandover(env, bgSharedGroupPtr, std::move(query), start, end, limit,columnIndices, ascending);
       } CATCH_STD()
       return 0;
@@ -1668,8 +1682,7 @@ JNIEXPORT jlong JNICALL Java_io_realm_internal_TableQuery_nativeImportHandoverTa
   {
     TR_ENTER_PTR(handoverPtr)
     SharedGroup::Handover<TableView> *handoverTableViewPtr = HO(TableView, handoverPtr);
-    std::unique_ptr<SharedGroup::Handover<TableView> > handoverTableView(handoverTableViewPtr);
-
+    std::unique_ptr<SharedGroup::Handover<TableView>> handoverTableView(handoverTableViewPtr);
     try {
         // import_from_handover will free (delete) the handover
         if (SG(callerSharedGrpPtr)->is_attached()) {

--- a/realm/realm-library/src/androidTest/java/io/realm/RealmAsyncQueryTests.java
+++ b/realm/realm-library/src/androidTest/java/io/realm/RealmAsyncQueryTests.java
@@ -18,6 +18,8 @@ package io.realm;
 
 import android.os.Handler;
 import android.os.SystemClock;
+import android.support.test.annotation.UiThreadTest;
+import android.support.test.rule.UiThreadTestRule;
 import android.support.test.runner.AndroidJUnit4;
 
 import org.junit.Rule;
@@ -32,12 +34,14 @@ import java.util.Random;
 import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.atomic.AtomicInteger;
 
+import io.realm.entities.AllJavaTypes;
 import io.realm.entities.AllTypes;
 import io.realm.entities.AnnotationIndexTypes;
 import io.realm.entities.Dog;
 import io.realm.entities.NonLatinFieldNames;
 import io.realm.entities.Owner;
 import io.realm.instrumentation.MockActivityManager;
+import io.realm.internal.async.RealmThreadPoolExecutor;
 import io.realm.internal.log.RealmLog;
 import io.realm.proxy.HandlerProxy;
 import io.realm.rule.RunInLooperThread;
@@ -59,6 +63,9 @@ public class RealmAsyncQueryTests {
     public final RunInLooperThread looperThread = new RunInLooperThread();
     @Rule
     public final TestRealmConfigurationFactory configFactory = new TestRealmConfigurationFactory();
+    @Rule
+    public final UiThreadTestRule uiThreadTestRule = new UiThreadTestRule();
+
 
     // ****************************
     // ****  Async transaction  ***
@@ -1809,6 +1816,136 @@ public class RealmAsyncQueryTests {
         }.start();
     }
 
+    // Test case for https://github.com/realm/realm-java/issues/2417
+    // Ensure that a UnreachableVersion exception during handover doesn't crash the app or cause a segfault.
+    @Test
+    @UiThreadTest
+    public void badVersion_findAll() throws NoSuchFieldException, IllegalAccessException {
+        TestHelper.replaceRealmThreadExectutor(RealmThreadPoolExecutor.newSingleThreadExecutor());
+        RealmConfiguration config  = configFactory.createConfiguration();
+        Realm realm = Realm.getInstance(config);
+        realm.executeTransactionAsync(new Realm.Transaction() {
+            @Override
+            public void execute(Realm realm) {
+                realm.deleteAll();
+            }
+        });
+        realm.executeTransactionAsync(new Realm.Transaction() {
+            @Override
+            public void execute(Realm realm) {
+                realm.deleteAll();
+            }
+        });
+        realm.executeTransactionAsync(new Realm.Transaction() {
+            @Override
+            public void execute(Realm realm) {
+                realm.deleteAll();
+            }
+        });
+        boolean result = realm.where(AllTypes.class).findAllAsync().load();
+        try {
+            assertFalse(result);
+        } finally {
+            realm.close();
+        }
+    }
+
+    // Test case for https://github.com/realm/realm-java/issues/2417
+    // Ensure that a UnreachableVersion exception during handover doesn't crash the app or cause a segfault.
+    @Test
+    @UiThreadTest
+    public void badVersion_findAllSortedAsync() throws NoSuchFieldException, IllegalAccessException {
+        TestHelper.replaceRealmThreadExectutor(RealmThreadPoolExecutor.newSingleThreadExecutor());
+        RealmConfiguration config = configFactory.createConfiguration();
+        Realm realm = Realm.getInstance(config);
+        realm.executeTransactionAsync(new Realm.Transaction() {
+            @Override
+            public void execute(Realm realm) {
+                realm.deleteAll();
+            }
+        });
+        realm.executeTransactionAsync(new Realm.Transaction() {
+            @Override
+            public void execute(Realm realm) {
+                realm.deleteAll();
+            }
+        });
+        realm.executeTransactionAsync(new Realm.Transaction() {
+            @Override
+            public void execute(Realm realm) {
+                realm.deleteAll();
+            }
+        });
+        realm.where(AllTypes.class)
+                .findAllSortedAsync(AllTypes.FIELD_STRING, Sort.ASCENDING, AllTypes.FIELD_LONG, Sort.DESCENDING)
+                .load();
+        realm.close();
+    }
+
+    // Test case for https://github.com/realm/realm-java/issues/2417
+    // Ensure that a UnreachableVersion exception during handover doesn't crash the app or cause a segfault.
+    @Test
+    @UiThreadTest
+    public void badVersion_distinct() throws NoSuchFieldException, IllegalAccessException {
+        TestHelper.replaceRealmThreadExectutor(RealmThreadPoolExecutor.newSingleThreadExecutor());
+        RealmConfiguration config = configFactory.createConfiguration();
+        Realm realm = Realm.getInstance(config);
+        realm.executeTransactionAsync(new Realm.Transaction() {
+            @Override
+            public void execute(Realm realm) {
+                realm.deleteAll();
+            }
+        });
+        realm.executeTransactionAsync(new Realm.Transaction() {
+            @Override
+            public void execute(Realm realm) {
+                realm.deleteAll();
+            }
+        });
+        realm.executeTransactionAsync(new Realm.Transaction() {
+            @Override
+            public void execute(Realm realm) {
+                realm.deleteAll();
+            }
+        });
+        realm.where(AllJavaTypes.class)
+                .distinctAsync(AllJavaTypes.FIELD_STRING)
+                .load();
+
+        realm.close();
+    }
+
+    // Test case for https://github.com/realm/realm-java/issues/2417
+    // Ensure that a UnreachableVersion exception during handover doesn't crash the app or cause a segfault.
+    @Test
+    @RunTestInLooperThread
+    public void badVersion_syncTransaction() throws NoSuchFieldException, IllegalAccessException {
+        TestHelper.replaceRealmThreadExectutor(RealmThreadPoolExecutor.newSingleThreadExecutor());
+        Realm realm = looperThread.realm;
+
+        // 1. Make sure that async query is not started
+        final RealmResults<AllTypes> result = realm.where(AllTypes.class).findAllSortedAsync(AllTypes.FIELD_STRING);
+        result.addChangeListener(new RealmChangeListener() {
+            @Override
+            public void onChange() {
+                // 4. The commit in #2, should result in a refresh being triggered, which means this callback will
+                // be notified once the updated async queries has run.
+                // with the correct
+                assertTrue(result.isValid());
+                assertTrue(result.isLoaded());
+                assertEquals(1, result.size());
+                looperThread.testComplete();
+            }
+        });
+
+        // 2. Advance the calle Realm, invalidating the version in the handover object
+        realm.beginTransaction();
+        realm.createObject(AllTypes.class);
+        realm.commitTransaction();
+
+        // 3. The async query should now (hopefully) fail with a BadVersion
+        result.load();
+    }
 
     // *** Helper methods ***
 

--- a/realm/realm-library/src/androidTest/java/io/realm/TestHelper.java
+++ b/realm/realm-library/src/androidTest/java/io/realm/TestHelper.java
@@ -33,6 +33,7 @@ import java.io.UnsupportedEncodingException;
 import java.lang.reflect.Constructor;
 import java.lang.reflect.Field;
 import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Modifier;
 import java.nio.charset.Charset;
 import java.security.MessageDigest;
 import java.security.NoSuchAlgorithmException;
@@ -48,6 +49,7 @@ import io.realm.entities.NullTypes;
 import io.realm.entities.StringOnly;
 import io.realm.internal.Table;
 import io.realm.internal.TableOrView;
+import io.realm.internal.async.RealmThreadPoolExecutor;
 import io.realm.internal.log.Logger;
 import io.realm.rule.TestRealmConfigurationFactory;
 
@@ -757,6 +759,20 @@ public class TestHelper {
         } catch (InvocationTargetException e) {
             throw new RuntimeException(e);
         }
+    }
+
+    /**
+     * Replaces the current thread executor with a another one for testing.
+     * WARNING: This method should only be called before any async tasks have been started.
+     *
+     * @param executor {@link RealmThreadPoolExecutor} that should replace the current one
+     */
+    public static RealmThreadPoolExecutor replaceRealmThreadExectutor(RealmThreadPoolExecutor executor) throws NoSuchFieldException, IllegalAccessException {
+        Field field = BaseRealm.class.getDeclaredField("asyncQueryExecutor");
+        field.setAccessible(true);
+        RealmThreadPoolExecutor oldExecutor = (RealmThreadPoolExecutor) field.get(null);
+        field.set(field, executor);
+        return oldExecutor;
     }
 
     /**

--- a/realm/realm-library/src/main/java/io/realm/BaseRealm.java
+++ b/realm/realm-library/src/main/java/io/realm/BaseRealm.java
@@ -56,7 +56,7 @@ abstract class BaseRealm implements Closeable {
     protected static final Map<Handler, String> handlers = new ConcurrentHashMap<Handler, String>();
 
     // Thread pool for all async operations (Query & transaction)
-    static final RealmThreadPoolExecutor asyncQueryExecutor = RealmThreadPoolExecutor.getInstance();
+    static final RealmThreadPoolExecutor asyncQueryExecutor = RealmThreadPoolExecutor.newDefaultExecutor();
 
     protected long threadId;
     protected RealmConfiguration configuration;

--- a/realm/realm-library/src/main/java/io/realm/HandlerController.java
+++ b/realm/realm-library/src/main/java/io/realm/HandlerController.java
@@ -34,6 +34,7 @@ import java.util.concurrent.Future;
 import io.realm.internal.IdentitySet;
 import io.realm.internal.Row;
 import io.realm.internal.SharedGroup;
+import io.realm.internal.async.BadVersionException;
 import io.realm.internal.async.QueryUpdateTask;
 import io.realm.internal.log.RealmLog;
 
@@ -95,31 +96,35 @@ public class HandlerController implements Handler.Callback {
         // that if a message does slip though (however unlikely), it will not try to update a SharedGroup that no
         // longer exists. `sharedGroupManager` will only be null if a Realm is really closed.
         if (realm.sharedGroupManager != null) {
+            QueryUpdateTask.Result result;
             switch (message.what) {
-                case REALM_CHANGED: {
+
+                case REALM_CHANGED:
                     realmChanged();
                     break;
-                }
-                case COMPLETED_ASYNC_REALM_RESULTS: {
-                    QueryUpdateTask.Result result = (QueryUpdateTask.Result) message.obj;
+
+                case COMPLETED_ASYNC_REALM_RESULTS:
+                    result = (QueryUpdateTask.Result) message.obj;
                     completedAsyncRealmResults(result);
                     break;
-                }
-                case COMPLETED_ASYNC_REALM_OBJECT: {
-                    QueryUpdateTask.Result result = (QueryUpdateTask.Result) message.obj;
+
+                case COMPLETED_ASYNC_REALM_OBJECT:
+                    result = (QueryUpdateTask.Result) message.obj;
                     completedAsyncRealmObject(result);
                     break;
-                }
-                case COMPLETED_UPDATE_ASYNC_QUERIES: {
+
+                case COMPLETED_UPDATE_ASYNC_QUERIES:
                     // this is called once the background thread completed the update of the async queries
-                    QueryUpdateTask.Result result = (QueryUpdateTask.Result) message.obj;
+                    result = (QueryUpdateTask.Result) message.obj;
                     completedAsyncQueriesUpdate(result);
                     break;
-                }
-                case REALM_ASYNC_BACKGROUND_EXCEPTION: {
+
+                case REALM_ASYNC_BACKGROUND_EXCEPTION:
                     // Don't fail silently in the background in case of Core exception
                     throw (Error) message.obj;
-                }
+
+                default:
+                    throw new IllegalArgumentException("Unknown message: " + message.what);
             }
         }
         return true;
@@ -437,7 +442,13 @@ public class HandlerController implements Handler.Callback {
                 // refresh the Realm to the version provided by the worker thread
                 // (advanceRead to the latest version may cause a version mismatch error) preventing us
                 // from importing correctly the handover table view
-                realm.sharedGroupManager.advanceRead(result.versionID);
+                try {
+                    realm.sharedGroupManager.advanceRead(result.versionID);
+                } catch (BadVersionException e) {
+                    // The version comparison above should have ensured that that the Caller version is less than the
+                    // Worker version. In that case it should always be safe to advance_read.
+                    throw new IllegalStateException("Failed to advance Caller Realm to Worker Realm version", e);
+                }
             }
 
             ArrayList<RealmResults<? extends RealmObject>> callbacksToNotify = new ArrayList<RealmResults<? extends RealmObject>>(result.updatedTableViews.size());
@@ -450,15 +461,15 @@ public class HandlerController implements Handler.Callback {
                     asyncRealmResults.remove(weakRealmResults);
 
                 } else {
+                    // update the instance with the new pointer
+                    realmResults.swapTableViewPointer(query.getValue());
+
                     // it's dangerous to notify the callback about new results before updating
                     // the pointers, because the callback may use another RealmResults not updated yet
                     // this is why we defer the notification until we're done updating all pointers
                     callbacksToNotify.add(realmResults);
 
                     RealmLog.d("COMPLETED_UPDATE_ASYNC_QUERIES realm:"+ HandlerController.this + " updating RealmResults " + weakRealmResults);
-
-                    // update the instance with the new pointer
-                    realmResults.swapTableViewPointer(query.getValue());
                 }
             }
 
@@ -501,8 +512,13 @@ public class HandlerController implements Handler.Callback {
                     // the caller has advanced we need to
                     // retry against the current version of the caller if it's still empty
                     if (realmObject.isValid()) { // already completed & has a valid pointer no need to re-run
+                        RealmLog.d("[COMPLETED_ASYNC_REALM_OBJECT "+ realmObject + "] , realm:" + HandlerController.this
+                                + " RealmObject is already loaded, just notify it.");
                         realmObject.notifyChangeListeners();
+
                     } else {
+                        RealmLog.d("[COMPLETED_ASYNC_REALM_OBJECT "+ realmObject + "] , realm:" + HandlerController.this
+                                + " RealmObject is not loaded yet. Rerun the query.");
                         RealmQuery<?> realmQuery = realmObjects.get(realmObjectWeakReference);
                         if (realmQuery == null) { // this is a retry of an empty RealmObject
                             realmQuery = emptyAsyncRealmObject.get(realmObjectWeakReference);
@@ -521,7 +537,7 @@ public class HandlerController implements Handler.Callback {
                 } else {
                     // should not happen, since the the background thread position itself against the provided version
                     // and the caller thread can only go forward (advance_read)
-                    throw new IllegalStateException("Caller thread behind the worker thread");
+                    throw new IllegalStateException("Caller thread behind the Worker thread");
                 }
             } // else: element GC'd in the meanwhile
         }

--- a/realm/realm-library/src/main/java/io/realm/RealmQuery.java
+++ b/realm/realm-library/src/main/java/io/realm/RealmQuery.java
@@ -36,6 +36,7 @@ import io.realm.internal.TableOrView;
 import io.realm.internal.TableQuery;
 import io.realm.internal.TableView;
 import io.realm.internal.async.ArgumentsHolder;
+import io.realm.internal.async.BadVersionException;
 import io.realm.internal.async.QueryUpdateTask;
 import io.realm.internal.log.RealmLog;
 
@@ -1492,6 +1493,11 @@ public class RealmQuery<E extends RealmObject> {
 
                         return handoverTableViewPointer;
 
+                    } catch (BadVersionException e) {
+                        // In some rare race conditions, this can happen. In that case, just ignore the error.
+                        RealmLog.d("findAllAsync handover could not complete due to a BadVersionException. " +
+                                "Retry is scheduled by a REALM_CHANGED event.");
+
                     } catch (Exception e) {
                         RealmLog.e(e.getMessage(), e);
                         closeSharedGroupAndSendMessageToHandler(sharedGroup,
@@ -1605,6 +1611,11 @@ public class RealmQuery<E extends RealmObject> {
                                 weakHandler, HandlerController.COMPLETED_ASYNC_REALM_RESULTS, result);
 
                         return handoverTableViewPointer;
+                    } catch (BadVersionException e) {
+                        // In some rare race conditions, this can happen. In that case, just ignore the error.
+                        RealmLog.d("findAllSortedAsync handover could not complete due to a BadVersionException. " +
+                                "Retry is scheduled by a REALM_CHANGED event.");
+
                     } catch (Exception e) {
                         RealmLog.e(e.getMessage(), e);
                         closeSharedGroupAndSendMessageToHandler(sharedGroup,
@@ -1771,6 +1782,11 @@ public class RealmQuery<E extends RealmObject> {
                                     weakHandler, HandlerController.COMPLETED_ASYNC_REALM_RESULTS, result);
 
                             return handoverTableViewPointer;
+                        } catch (BadVersionException e) {
+                            // In some rare race conditions, this can happen. In that case, just ignore the error.
+                            RealmLog.d("findAllSortedAsync handover could not complete due to a BadVersionException. " +
+                                    "Retry is scheduled by a REALM_CHANGED event.");
+
                         } catch (Exception e) {
                             RealmLog.e(e.getMessage(), e);
                             closeSharedGroupAndSendMessageToHandler(sharedGroup,

--- a/realm/realm-library/src/main/java/io/realm/RealmResults.java
+++ b/realm/realm-library/src/main/java/io/realm/RealmResults.java
@@ -34,6 +34,7 @@ import io.realm.internal.TableOrView;
 import io.realm.internal.TableQuery;
 import io.realm.internal.TableView;
 import io.realm.internal.Table;
+import io.realm.internal.async.BadVersionException;
 import io.realm.internal.log.RealmLog;
 import rx.Observable;
 
@@ -784,7 +785,6 @@ public final class RealmResults<E extends RealmObject> extends AbstractList<E> {
             throw new RealmException("Replacing elements not supported.");
         }
 
-
         /**
          * Removes the RealmObject at the current position from both the list and the underlying Realm.
          *
@@ -800,10 +800,15 @@ public final class RealmResults<E extends RealmObject> extends AbstractList<E> {
      * thread.
      *
      * @param handoverTableViewPointer handover pointer to the new table_view.
+     * @throws IllegalStateException if caller and worker are not at the same version.
      */
     void swapTableViewPointer(long handoverTableViewPointer) {
-        table = query.importHandoverTableView(handoverTableViewPointer, realm.sharedGroupManager.getNativePointer());
-        isCompleted = true;
+        try {
+            table = query.importHandoverTableView(handoverTableViewPointer, realm.sharedGroupManager.getNativePointer());
+            isCompleted = true;
+        } catch (BadVersionException e) {
+            throw new IllegalStateException("Caller and Worker Realm should have been at the same version");
+        }
     }
 
     /**

--- a/realm/realm-library/src/main/java/io/realm/internal/ImplicitTransaction.java
+++ b/realm/realm-library/src/main/java/io/realm/internal/ImplicitTransaction.java
@@ -16,6 +16,8 @@
 
 package io.realm.internal;
 
+import io.realm.internal.async.BadVersionException;
+
 public class ImplicitTransaction extends Group {
 
     private final SharedGroup parent;
@@ -38,7 +40,7 @@ public class ImplicitTransaction extends Group {
      *
      * @param versionID version of the shared group.
      */
-    public void advanceRead(SharedGroup.VersionID versionID) {
+    public void advanceRead(SharedGroup.VersionID versionID) throws BadVersionException {
         assertNotClosed();
         parent.advanceRead(versionID);
     }

--- a/realm/realm-library/src/main/java/io/realm/internal/SharedGroup.java
+++ b/realm/realm-library/src/main/java/io/realm/internal/SharedGroup.java
@@ -20,6 +20,7 @@ import java.io.Closeable;
 import java.io.IOError;
 
 import io.realm.exceptions.RealmIOException;
+import io.realm.internal.async.BadVersionException;
 
 public class SharedGroup implements Closeable {
 
@@ -82,7 +83,7 @@ public class SharedGroup implements Closeable {
         nativeAdvanceRead(nativePtr);
     }
 
-    void advanceRead(VersionID versionID) {
+    void advanceRead(VersionID versionID) throws BadVersionException {
         nativeAdvanceReadToVersion(nativePtr, versionID.version, versionID.index);
     }
 
@@ -317,6 +318,6 @@ public class SharedGroup implements Closeable {
     private native void nativeRollbackAndContinueAsRead(long nativePtr);
     private native long[] nativeGetVersionID (long nativePtr);
     private native void nativeAdvanceRead(long nativePtr);
-    private native void nativeAdvanceReadToVersion(long nativePtr, long version, long index);
+    private native void nativeAdvanceReadToVersion(long nativePtr, long version, long index) throws BadVersionException;
     private native void nativePromoteToWrite(long nativePtr);
 }

--- a/realm/realm-library/src/main/java/io/realm/internal/SharedGroupManager.java
+++ b/realm/realm-library/src/main/java/io/realm/internal/SharedGroupManager.java
@@ -21,6 +21,7 @@ import java.io.File;
 import java.io.IOException;
 
 import io.realm.RealmConfiguration;
+import io.realm.internal.async.BadVersionException;
 import io.realm.internal.log.RealmLog;
 
 /**
@@ -78,7 +79,7 @@ public class SharedGroupManager implements Closeable {
     /**
      * Advances the Realm file to the given version.
      */
-    public void advanceRead(SharedGroup.VersionID version) {
+    public void advanceRead(SharedGroup.VersionID version) throws BadVersionException {
         transaction.advanceRead(version);
     }
 

--- a/realm/realm-library/src/main/java/io/realm/internal/TableQuery.java
+++ b/realm/realm-library/src/main/java/io/realm/internal/TableQuery.java
@@ -21,6 +21,7 @@ import java.util.Date;
 
 import io.realm.Case;
 import io.realm.Sort;
+import io.realm.internal.async.BadVersionException;
 
 public class TableQuery implements Closeable {
     protected boolean DEBUG = false;
@@ -473,28 +474,28 @@ public class TableQuery implements Closeable {
     // handover find* methods
     // this will use a background SharedGroup to import the query (using the handover object)
     // run the query, and return the table view to the caller SharedGroup using the handover object.
-    public long findAllWithHandover(long bgSharedGroupPtr, long nativeReplicationPtr,  long ptrQuery) {
+    public long findAllWithHandover(long bgSharedGroupPtr, long nativeReplicationPtr,  long ptrQuery) throws BadVersionException {
         validateQuery();
         // Execute the disposal of abandoned realm objects each time a new realm object is created
         context.executeDelayedDisposal();
         return nativeFindAllWithHandover(bgSharedGroupPtr, ptrQuery, 0, Table.INFINITE, Table.INFINITE);
     }
 
-    public long findDistinctWithHandover(long bgSharedGroupPtr, long nativeReplicationPtr,  long ptrQuery, long columnIndex) {
+    public long findDistinctWithHandover(long bgSharedGroupPtr, long nativeReplicationPtr,  long ptrQuery, long columnIndex) throws BadVersionException {
         validateQuery();
         // Execute the disposal of abandoned realm objects each time a new realm object is created
         context.executeDelayedDisposal();
         return nativeGetDistinctViewWithHandover(bgSharedGroupPtr, ptrQuery, columnIndex);
     }
 
-    public long findAllSortedWithHandover(long bgSharedGroupPtr, long nativeReplicationPtr, long ptrQuery, long columnIndex, Sort sortOrder) {
+    public long findAllSortedWithHandover(long bgSharedGroupPtr, long nativeReplicationPtr, long ptrQuery, long columnIndex, Sort sortOrder) throws BadVersionException {
         validateQuery();
         // Execute the disposal of abandoned realm objects each time a new realm object is created
         context.executeDelayedDisposal();
         return nativeFindAllSortedWithHandover(bgSharedGroupPtr, ptrQuery, 0, Table.INFINITE, Table.INFINITE, columnIndex, sortOrder.getValue());
     }
 
-    public long findAllMultiSortedWithHandover(long bgSharedGroupPtr, long nativeReplicationPtr, long ptrQuery, long[] columnIndices, Sort[] sortOrders) {
+    public long findAllMultiSortedWithHandover(long bgSharedGroupPtr, long nativeReplicationPtr, long ptrQuery, long[] columnIndices, Sort[] sortOrders) throws BadVersionException {
         validateQuery();
         // Execute the disposal of abandoned realm objects each time a new realm object is created
         context.executeDelayedDisposal();
@@ -502,8 +503,15 @@ public class TableQuery implements Closeable {
         return nativeFindAllMultiSortedWithHandover(bgSharedGroupPtr, ptrQuery, 0, Table.INFINITE, Table.INFINITE, columnIndices, ascendings);
     }
 
-    // Suppose to be called from the caller SharedGroup thread
-    public TableView importHandoverTableView(long handoverPtr, long callerSharedGroupPtr) {
+    /**
+     * Imports a TableView from a worker thread to the caller thread.
+     *
+     * @param handoverPtr pointer to the handover object
+     * @param callerSharedGroupPtr pointer to the SharedGroup on the caller thread.
+     * @return the TableView on the caller thread.
+     * @throws BadVersionException if the worker thread and caller thread are not at the same version.
+     */
+    public TableView importHandoverTableView(long handoverPtr, long callerSharedGroupPtr) throws BadVersionException {
         long nativeTvPtr = nativeImportHandoverTableViewIntoSharedGroup(handoverPtr, callerSharedGroupPtr);
         try {
             return new TableView(this.context, this.table, nativeTvPtr);
@@ -797,14 +805,14 @@ public class TableQuery implements Closeable {
     private native void nativeIsNotNull(long nativePtr, long columnIndices[]);
     private native long nativeCount(long nativeQueryPtr, long start, long end, long limit);
     private native long nativeRemove(long nativeQueryPtr, long start, long end, long limit);
-    private native long nativeImportHandoverTableViewIntoSharedGroup(long handoverTableViewPtr, long callerSharedGroupPtr);
+    private native long nativeImportHandoverTableViewIntoSharedGroup(long handoverTableViewPtr, long callerSharedGroupPtr) throws BadVersionException;
     private native long nativeHandoverQuery(long callerSharedGroupPtr, long nativeQueryPtr);
-    public static native long nativeFindAllSortedWithHandover(long bgSharedGroupPtr, long nativeQueryPtr, long start, long end, long limit, long columnIndex, boolean ascending);
-    public static native long nativeFindAllWithHandover(long bgSharedGroupPtr, long nativeQueryPtr, long start, long end, long limit);
-    public static native long nativeGetDistinctViewWithHandover(long bgSharedGroupPtr, long nativeQueryPtr, long columnIndex);
+    public static native long nativeFindAllSortedWithHandover(long bgSharedGroupPtr, long nativeQueryPtr, long start, long end, long limit, long columnIndex, boolean ascending) throws BadVersionException;
+    public static native long nativeFindAllWithHandover(long bgSharedGroupPtr, long nativeQueryPtr, long start, long end, long limit) throws BadVersionException;
+    public static native long nativeGetDistinctViewWithHandover(long bgSharedGroupPtr, long nativeQueryPtr, long columnIndex) throws BadVersionException;
     public static native long nativeFindWithHandover(long bgSharedGroupPtr, long nativeQueryPtr, long fromTableRow);
-    public static native long nativeFindAllMultiSortedWithHandover(long bgSharedGroupPtr, long nativeQueryPtr, long start, long end, long limit, long[] columnIndices, boolean[] ascending);
+    public static native long nativeFindAllMultiSortedWithHandover(long bgSharedGroupPtr, long nativeQueryPtr, long start, long end, long limit, long[] columnIndices, boolean[] ascending) throws BadVersionException;
     public static native long nativeImportHandoverRowIntoSharedGroup(long handoverRowPtr, long callerSharedGroupPtr);
     public static native void nativeCloseQueryHandover(long nativePtr);
-    public static native long[] nativeBatchUpdateQueries(long bgSharedGroupPtr, long[] handoverQueries, long[][] parameters, long[][] queriesParameters, boolean[][] multiSortOrder);
+    public static native long[] nativeBatchUpdateQueries(long bgSharedGroupPtr, long[] handoverQueries, long[][] parameters, long[][] queriesParameters, boolean[][] multiSortOrder) throws BadVersionException;
 }

--- a/realm/realm-library/src/main/java/io/realm/internal/async/BadVersionException.java
+++ b/realm/realm-library/src/main/java/io/realm/internal/async/BadVersionException.java
@@ -16,7 +16,6 @@
 
 package io.realm.internal.async;
 
-import io.realm.exceptions.RealmException;
 import io.realm.internal.Keep;
 
 /**
@@ -24,7 +23,7 @@ import io.realm.internal.Keep;
  * state of the Realm which might be more up-to-date than the provided results or vice versa.
  */
 @Keep
-public class BadVersionException extends RealmException {
+public class BadVersionException extends Exception {
 
     public BadVersionException(String detailMessage) {
         super(detailMessage);

--- a/realm/realm-library/src/main/java/io/realm/internal/async/RealmThreadPoolExecutor.java
+++ b/realm/realm-library/src/main/java/io/realm/internal/async/RealmThreadPoolExecutor.java
@@ -39,21 +39,22 @@ public class RealmThreadPoolExecutor extends ThreadPoolExecutor {
     private ReentrantLock pauseLock = new ReentrantLock();
     private Condition unpaused = pauseLock.newCondition();
 
-    private static volatile RealmThreadPoolExecutor instance;
-
-    public static RealmThreadPoolExecutor getInstance() {
-           if (instance == null) {
-               synchronized (RealmThreadPoolExecutor.class) {
-                   if (instance == null) {
-                       instance = new RealmThreadPoolExecutor();
-                   }
-               }
-           }
-        return instance;
+    /**
+     * Creates a default RealmThreadPool that is bounded by the number of available cores.
+     */
+    public static RealmThreadPoolExecutor newDefaultExecutor() {
+        return new RealmThreadPoolExecutor(CORE_POOL_SIZE, CORE_POOL_SIZE);
     }
 
-    private RealmThreadPoolExecutor() {
-        super(CORE_POOL_SIZE, CORE_POOL_SIZE,
+    /**
+     * Creates a RealmThreadPool with only 1 thread. This is primarily useful for testing.
+     */
+    public static RealmThreadPoolExecutor newSingleThreadExecutor() {
+        return new RealmThreadPoolExecutor(1, 1);
+    }
+
+    private RealmThreadPoolExecutor(int corePoolSize, int maxPoolSize) {
+        super(corePoolSize, maxPoolSize,
                 0L, TimeUnit.MILLISECONDS, //terminated idle thread
                 new ArrayBlockingQueue<Runnable>(QUEUE_SIZE));
     }

